cobbler/api.py

```diff
@@ -1076,6 +1076,45 @@ def new_menu(self, is_subobject: bool = False, **kwargs: Any) -> "menu.Menu":
         return menu.Menu(self, is_subobject, **kwargs)
 
     # ==========================================================================
+    def add_remove_items(self, items: List[Tuple[str, "InheritableItem", bool, float, str]]):
+        """
+        Add or remove multiple items.
+
+        :param items: The list of items.
+        """
+        for what, ref, _, mtime, base_id in items:
+            if base_id:
+                orig_ref = self.find_items(what, criteria={"uid": base_id}, return_list=False)
+                if not orig_ref or isinstance(orig_ref, list) or mtime != orig_ref.mtime:
+                    raise ValueError("The item has been modified")
+
+        to_remove = [i for i in items if i[2]]
+        to_add = [i for i in items if not i[2]]
+
+        to_remove.sort(key=lambda x: -x[1].depth)
+        to_add.sort(key=lambda x: x[1].depth)
```

Comment:
We cannot do this as we would destroy the logical call order. Ensuring this is the clients responsibility. Otherwise, the behaviour with and without a transaction starts to differ and I have no desire to debug stuff like this.

Reply:
Now the transaction is not implemented as a list of operations that is replayed. It is a collection of items in their final state. If an item is modified multiple times, this list contains only one entry, which needs to be saved.

The sorting ensures that the items are deleted from children to parents and created from parents to children.